packages/arb-validator/checkpointing/indexedCheckpointer.go

```diff
@@ -0,0 +1,477 @@
+/*
+* Copyright 2020, Offchain Labs, Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+ */
+
+package checkpointing
+
+import (
+	"context"
+	"errors"
+	"github.com/golang/protobuf/proto"
+	"github.com/offchainlabs/arbitrum/packages/arb-avm-cpp/cmachine"
+	"github.com/offchainlabs/arbitrum/packages/arb-util/common"
+	"github.com/offchainlabs/arbitrum/packages/arb-util/machine"
+	"github.com/offchainlabs/arbitrum/packages/arb-util/value"
+	"github.com/offchainlabs/arbitrum/packages/arb-validator/arbbridge"
+	"log"
+	"math/big"
+	"os"
+	"sync"
+	"time"
+)
+
+type IndexedCheckpointer struct {
+	*sync.Mutex
+	db                    machine.CheckpointStorage
+	maxReorgHeight        *big.Int
+	nextCheckpointToWrite *writableCheckpoint
+	chansToClose          []chan struct{}
+}
+
+func NewIndexedCheckpointerFactory(
+	rollupAddr common.Address,
+	arbitrumCodeFilePath string,
+	databasePath string,
+	maxReorgHeight *big.Int,
+	forceFreshStart bool,
+) RollupCheckpointerFactory {
+	if databasePath == "" {
+		databasePath = MakeCheckpointDatabasePath(rollupAddr)
+	}
+	if forceFreshStart {
+		// for testing only --  delete old database to get fresh start
+		if err := os.RemoveAll(databasePath); err != nil {
+			log.Fatal(err)
+		}
+	}
+	cCheckpointer, err := cmachine.NewCheckpoint(databasePath, arbitrumCodeFilePath)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	ret := &IndexedCheckpointer{
+		new(sync.Mutex),
+		cCheckpointer,
+		new(big.Int).Set(maxReorgHeight),
+		nil,
+		nil,
+	}
+	go ret.writeDaemon()
+	go ret.cleanupDaemon()
+	return ret
+}
+
+func (cp *IndexedCheckpointer) New(_ context.Context) RollupCheckpointer {
```

Comment:
This method is pretty strange looking. Is there a reason for it?

Reply:
Yes.  This code has to satisfy an interface that uses a factory pattern.  The interface does this so it can kill the checkpointer on a reorg, then call the factory again to make a new checkpointer to use going forward.

But this code doesn't do a kill/restart on reorg. This code is reorg-aware, so it just make a single instance once and returns the same one every time the factory is called.  And rather than having a separate factory object, I just made a single object that acts as its own factory, returning itself.

Added a comment to the source code explaining this.